### PR TITLE
[mesheryctl] Make 'system reset' context aware

### LINKED ISSUE #2248: [mesheryctl] make `system reset` context-aware
#### Current Behavior
`mesheryctl system reset` incidentally crosses release channels by reaching out to GitHub and pulling the latest `docker-compose.yaml`.

#### Desired Behavior
`reset` needs to consider for the current-context and the `channel` and `version` attributes when pulling from GitHub.

#### Implementation
1. No behavior change for `edge` channel. Just grab latest from GitHub `master` branch.
1. For `stable` channel:
   1.  if version is specified grab the the `docker-compose.yaml` from the git history at that release tag.
   1.  if version is “latest”, then reuse `version` for checking for latest stable release tag and then pull the docker-compose.yaml based on that tag.

### PR DESCRIPTION
**Pull request by pottekkat**
Signed-off-by: navendu-pottekkat <navendupottekkat@gmail.com>

**Description**

This PR fixes #2248 

**Notes for Reviewers**

This is a draft PR and is not ready for review.

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by anirudhjain75** [comment]
This is still a WIP do ping me when you're partially or completely done. Or if you need my help. As far as I can see the changes are yet to be implemented

**Comment by navendu-pottekkat** [comment]
@anirudhjain75 @leecalcote Yes, this is a WIP. I still have to implement the fetch from GitHub using the API part. I will try to get it done today.

**Comment by navendu-pottekkat** [comment]
@leecalcote @anirudhjain75 I have made a basic fix for this issue that seem to be working while I tested locally. I still need to setup some error handling for all possible scenarios. But this works and please check it and let me know for any changes.

**Comment by leecalcote** [comment]
@navendu-pottekkat, I don't believe that this was part of the original ask, but will you consider support of the `-c` flag for temporarily overriding the current context? Support for this at the `system` level is coming. @anirudhjain75 does any consideration need to be made here in order to support `-c` as written here - https://github.com/layer5io/meshery/pull/2256/files#diff-3ec6eb99922674fe5326ef77839a41f6e5356441ffebb8501c2196ace27a93aaR71?

I assume not given that the `system` level support should provide this to all `system` level commands.

File: mesheryctl/internal/cli/root/system/reset.go [hunk 1]
```diff
@@ -39,6 +41,29 @@ var resetCmd = &cobra.Command{
...
+	if currChannel == "edge" {
+		fileURL = "https://raw.githubusercontent.com/layer5io/meshery/master/docker-compose.yaml"
+	} else if currChannel == "stable" {
+		fileURL = "https://raw.githubusercontent.com/layer5io/meshery/" + currVersion + "/docker-compose.yaml"
+	}
+
+	log.Printf("Current Context: %s", currentContext)
+	log.Printf("Current Channel: %s", currChannel)
+	log.Printf("Current Version: %s", currVersion)
+	log.Printf("File URL: %s", fileURL)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/reset.go:65]: @navendu-pottekkat, yes, to your question, this debug line can be removed.

File: mesheryctl/internal/cli/root/system/reset.go [hunk 2]
```diff
@@ -39,6 +41,29 @@ var resetCmd = &cobra.Command{
...
 func resetMesheryConfig() error {
 	log.Info("Meshery resetting...")
+
+	mctlCfg, err := config.GetMesheryCtl(viper.GetViper())
+	if err != nil {
+		return errors.Wrap(err, "error processing config")
+	}
+
+	currentContext := mctlCfg.CurrentContext
+	currChannel := mctlCfg.Contexts[currentContext].Channel
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/reset.go:51]: The `system` level struct should be filtering any invalid values for channel and version, so confirmation of valid values shouldn't have to be considered here.

@anirudhjain75 will you confirm that this is the case?

**Review by leecalcote — DISMISSED** [review]

File: mesheryctl/internal/cli/root/system/reset.go (hunk 1 above)
Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/reset.go:65]: @leecalcote What should it output instead?

**Comment by navendu-pottekkat** [comment]
> @navendu-pottekkat, I don't believe that this was part of the original ask, but will you consider support of the `-c` flag for temporarily overriding the current context? Support for this at the `system` level is coming. @anirudhjain75 does any consideration need to be made here in order to support `-c` as written here - https://github.com/layer5io/meshery/pull/2256/files#diff-3ec6eb99922674fe5326ef77839a41f6e5356441ffebb8501c2196ace27a93aaR71?
> 
> I assume not given that the `system` level support should provide this to all `system` level commands.

I think it is better to raise this as a separate issue and make a new PR

**Comment by navendu-pottekkat** [comment]
@anirudhjain75 @leecalcote I have made the changes to the logs. It looks like this now-

```
navendu@navendu:~/Code/meshery-dev/meshery/mesheryctl$ ./mesheryctl system reset
Meshery resetting...

Current Context: local
Channel: stable
Version: v0.4.20

Fetching default docker-compose file at version: v0.4.20...

...Meshery config (/home/navendu/.meshery/meshery.yaml) now reset to default settings.
```

Overriding the context using the -c flag can be made into another issue and a separate PR can be made.

@anirudhjain75 Please review and merge this PR.

image

**Review by hexxdump — COMMENTED** [review]
instead of default settings - can we print currVersion ? or default settings of currVersion?

**Comment by navendu-pottekkat** [comment]
> instead of default settings - can we print currVersion ? or default settings of currVersion?

We are actually printing the version. Here v0.4.20 is the version.
```
Fetching default docker-compose file at version: v0.4.20...

...Meshery config (/home/navendu/.meshery/meshery.yaml) now reset to default settings.
```

**Comment by leecalcote** [comment]
> > @navendu-pottekkat, I don't believe that this was part of the original ask, but will you consider support of the `-c` flag for temporarily overriding the current context? Support for this at the `system` level is coming. @anirudhjain75 does any consideration need to be made here in order to support `-c` as written here - https://github.com/layer5io/meshery/pull/2256/files#diff-3ec6eb99922674fe5326ef77839a41f6e5356441ffebb8501c2196ace27a93aaR71?
> > I assume not given that the `system` level support should provide this to all `system` level commands.
> 
> I think it is better to raise this as a separate issue and make a new PR

Sounds real good. Let's do a new issue/PR.

**Review by leecalcote — APPROVED** [review]
Looking good. A follow up issue on "platform" support for "docker" and for `-c` flag...

**Pull request merged**